Teuben gbt

teuben opened the pull request:
2 issues resolved, though the units fix seems really ugly.
The 3rd issue about the all() boolean iteration is a mystery to me since it works in an ipython session. The test_3c284.py will trigger this error, though my PR has renamed the range= to srange= to avoid colliding with range(). Bravo emacs color highlighighter.

teuben committed:
Fixed two bitrot items, and some minor docs updated
1) unit conversions such that
       sp.xarr -= delta_freq
   would work, even though after all this abuse that line had to become
       sp.xarr = sp.xarr - delta_freq
2) calOK = (calON - (onoff=='OFF')) does not work in modern python, this is now
   calOK = np.bitwise_xor(calON, onoff=='OFF')

Added some documentation

teuben committed:
Add some @todo's about the various SDFITS dialects ?

teuben committed:
Unresolved problem using all() showing up via test_3c284.py

First of all the range= keyword would hide the internal range() command, so I renamed it,
but my problem wasn't a side-effect of this. I did a bunch of stupid expansions to get to the bottom of
this, but it's persistently failing with TypeError: 'bool' object is not iterable

If I cut and paste the whole mess into an ipython session, it all works, the
original line
    if not (spectrum1.xarr.shape == spectrum2.xarr.shape) or not all(spectrum1.xarr == spectrum2.xarr):
just works. But running in the code, not. Is this one of those 1+1=3 cases?

keflavich commented on pyspeckit/spectrum/readers/gbt.py:
```diff
@@ -98,9 +117,17 @@ def read_gbt_scan(sdfitsfile, obsnumber=0):
+    # some creative abuse how a S.A. stores and converts units, just in order to convert VFRAME in m/s to a delta_freq in Hz
+    c1 = pyspeckit.spectrum.units.SpectroscopicAxis(obsfreq,'Hz',velocity_convention='radio',
+                                                    refX=obsfreq, refX_unit=sp.xarr.unit)
+    v1 = (c1.as_unit('m/s') + u.Quantity(header['VFRAME'],'m/s')).value
```
I think this section should just be simplified.  This is a lot of complexity that is not justified.
```
centerfreq = u.Quantity(obsfreq, u.Hz)
delta_freq = u.Quantity(header['VFRAME'], u.m/u.s).to(u.Hz, u.doppler_radio(centerfreq)) - centerfreq
sp.xarr = sp.xarr - delta_freq
```

keflavich commented on pyspeckit/spectrum/readers/gbt.py:
```diff
@@ -129,7 +156,7 @@ def read_gbt_target(sdfitsfile, objectname, verbose=False):
             whnod = bintable.data['PROCSEQN'] == nod
             for onoff in ('ON','OFF'):
-                calOK = (calON - (onoff=='OFF'))
+                calOK = np.bitwise_xor(calON, onoff=='OFF')
```
I wonder if this was actually supposed to be `calOK = calON & not (onoff == 'OFF')` ?  I don't see why `xor` logic should be applied here... even though that is a correct interpretation of `-` in this case.

keflavich commented on pyspeckit/spectrum/readers/gbt.py:
```diff
@@ -180,18 +207,18 @@ def reduce_nod(blocks, verbose=False, average=True, fdid=(1,2)):
     # for each pair...
     for sampname in nodpairs:
-        on1 = sampname+"ON1"
+        on1  = sampname+"ON1"
```
don't add extra whitespace here

teuben committed:
why did this ever work?  This caused python3 to complain
   The truth value of a Quantity is ambiguous. In the future this will raise a ValueError.

teuben committed:
lots of horrific hacks to make it work, meant for reduce_ps() for NGC5291
